pandora-common/src/gameLogic/character/character.ts

```diff
@@ -15,14 +16,16 @@ export abstract class GameLogicCharacter extends TypedEventEmitter<GameLogicChar
 	public readonly id: CharacterId;
 	public readonly accountId: AccountId;
 	public readonly name: string;
+	public readonly publicData: Immutable<ICharacterPublicData>;
```

Comment:
This is problematic. Here you have a `readonly` property that is marked as `Immutable`, while in fact you expect it to be updated elsewhere. I really dislike such pattern. It also relies heavily on multiple places having reference to a same object, which is an invariant that is easy to break.

Instead you should look into how the `InteractionSubsystem` is done - this character really should only have the minimal data (which is special because it _never_ changes after character creation is finished).
The asset preferences should then be added as `abstract readonly` in this class, an updatable property in the `GameLogicCharacterClient` and either an updatable property or getter (a function passed through constructor) to the `GameLogicCharacterServer` class.

Reply:
other then this I fixed the rest, will look at this tomorrow

Reply:
well I looked into this, but wherever I insert a getter and or a subscription, it just feels like pure bloat that does nothing but add more code with extra complexity and more things to do for the server while being completely useless

if we somehow need to track the event when the change occurs then I can see this as justifiable but as it currently stands I do not see this as a potential improvement

`Immutable` is nothing more than a deep readonly, it doesn't mean a persistent or frozen object, in this case, it is just a constant reference

JS references are the simplest way to have a shared state, and the Immutable ensures that from `GameLogicCharacter` cannot modify it, I would not throw out the simplicity of using a reference unless using reference is not possible

we can just rename the variable to `publicDataRef` or something and be done with it